zkml/src/layers/dense.rs

```diff
@@ -123,6 +124,255 @@ impl<T: Number> Dense<T> {
+    }
+}
+
+impl<N: Number> Evaluate<N> for Dense<N> {
+    fn evaluate<E: ExtensionField>(
+        &self,
+        inputs: &[&Tensor<N>],
+        _unpadded_input_shapes: Vec<Vec<usize>>,
+    ) -> Result<LayerOut<N, E>, ProvableOpError> {
+        if inputs.len() != 1 {
+            return Err(ProvableOpError::ParameterError(
+                "Dense layer expects one input".to_string(),
+            ));
+        }
+        let input = inputs[0];
+        Ok(LayerOut::from_vec(vec![if input.get_shape().len() != 1 {
+            let flat_input = input.flatten();
+            let matvec = self.matrix.matvec(&flat_input);
+            matvec.add(&self.bias)
+        } else {
+            self.matrix.matvec(input).add(&self.bias)
+        }]))
+    }
+}
+
+impl<E> ProveInfo<E> for Dense<Element>
+where
+    E: ExtensionField,
+    E::BaseField: Serialize + DeserializeOwned,
+    E: Serialize + DeserializeOwned,
+{
+    fn step_info(
+        &self,
+        id: PolyID,
+        mut aux: ContextAux,
+    ) -> Result<(LayerCtx<E>, ContextAux), ProvableOpError> {
+        // construct dimension of the polynomial given to the sumcheck
+        let ncols = self.matrix.ncols_2d();
+        aux.last_output_shape
+            .iter_mut()
+            .for_each(|shape| *shape = vec![self.matrix.nrows_2d()]);
+        // each poly is only two polynomial right now: matrix and vector
+        // for matrix, each time we fix the variables related to rows so we are only left
+        // with the variables related to columns
+        let matrix_num_vars = ncols.ilog2() as usize;
+        let vector_num_vars = matrix_num_vars;
+        // there is only one product (i.e. quadratic sumcheck)
+        let dense_info = LayerCtx::Dense(DenseCtx {
+            matrix_poly_id: id,
+            matrix_poly_aux: VPAuxInfo::<E>::from_mle_list_dimensions(&vec![vec![
+                matrix_num_vars,
+                vector_num_vars,
+            ]]),
+            bias_poly_id: BIAS_POLY_ID + id,
+            unpadded_matrix_shape: self.unpadded_matrix_shape.clone(),
+            padded_matrix_shape: self.matrix.get_shape().to_vec(),
+        });
+        Ok((dense_info, aux))
+    }
+
+    fn commit_info(&self, id: NodeId) -> Vec<Option<(PolyID, Vec<E>)>> {
+        let evals = self.matrix.evals_2d();
+        let bias_evals = self.bias.evals_flat();
+        let id = id as PolyID;
+        debug!(
+            "Commitment : dense layer ID {}: size {}",
+            id,
+            evals.len().ilog2()
+        );
+        debug!(
+            "Commitment : dense layer bias ID {}: size {}",
+            BIAS_POLY_ID + id,
+            bias_evals.len().ilog2()
+        );
+        vec![Some((id, evals)), Some((BIAS_POLY_ID + id, bias_evals))]
+    }
+}
+
+impl PadOp for Dense<Element> {
+    fn pad_node(self, si: &mut ShapeInfo) -> Result<Self, ProvableOpError>
+    where
+        Self: Sized,
+    {
+        pad_dense(self, si).map_err(|e| ProvableOpError::GenericError(e))
```

Comment:
Either we should have a PaddingError OR we should get rid of our own error. personally i'm not  strong on having our own error struct, as we rarely "retry" depending on one or the other error. @Zyouell was keen on this but if so, we should go all the way.